dockerfiles/Dockerfile.training

```diff
@@ -0,0 +1,200 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# multi-stage arguments (repeat ARG NAME below)
+ARG UCX_VERSION=1.8.0
+ARG OPENMPI_VERSION=4.0.4
+ARG CONDA_VERSION=4.7.10
+ARG NUMPY_VERSION=1.18.5
+ARG ONNX_VERSION=1.7.0
+ARG PYTORCH_VERSION=1.6.0
+
+# cuda development image for building sources
+FROM nvidia/cuda:10.2-cudnn7-devel-ubuntu18.04 as build
+
+# set location for builds
+WORKDIR /stage
+
+# install curl and git
+RUN apt-get -y update &&\
+    apt-get -y --no-install-recommends install \
+        curl \
+        git
+    
+# install miniconda (comes with python 3.7 default)
+ARG CONDA_VERSION
+ARG CONDA_URL=https://repo.anaconda.com/miniconda/Miniconda3-${CONDA_VERSION}-Linux-x86_64.sh
+RUN cd /stage && curl -fSsL --insecure ${CONDA_URL} -o install-conda.sh &&\
+    /bin/bash ./install-conda.sh -b -p /opt/conda &&\
+    /opt/conda/bin/conda clean -ya
+ENV PATH=/opt/conda/bin:${PATH}
+
+# install cmake, setuptools, numpy, and onnx
+ARG NUMPY_VERSION
+ARG ONNX_VERSION
+RUN conda install -y \
+        setuptools \
+        cmake \
+        numpy=${NUMPY_VERSION} &&\
+    pip install \
+        onnx=="${ONNX_VERSION}"
+
+# build ucx suite
+# note: openmpi will not select ucx without multithreading enabled
+ARG UCX_VERSION
+ARG UCX_TARNAME=ucx-$UCX_VERSION
+ARG UCX_URL=https://github.com/openucx/ucx/releases/download/v${UCX_VERSION}/${UCX_TARNAME}.tar.gz
+RUN apt-get -y update && apt-get -y --no-install-recommends install \
+        libibverbs-dev \
+        libnuma-dev &&\
+    cd /stage && curl -fSsL ${UCX_URL} | tar xzf - &&\
+    cd ${UCX_TARNAME} &&\
+    ./configure \
+	--prefix=/opt/ucx \
+        --with-cuda=/usr/local/cuda \
+        --with-verbs=/usr/lib/x86_64-linux-gnu \
+        --enable-mt &&\
+    make -j"$(nproc)" &&\
+    make install
+
+# build openmpi (use --prefix /opt/openmpi-xxx to move to runtime image)
+# note: require --enable-orterun-prefix-by-default for Azure machine learning compute
+# note: disable verbs as we use ucx middleware and don't want btl openib warnings
+ARG OPENMPI_VERSION
+ARG OPENMPI_TARNAME=openmpi-${OPENMPI_VERSION}
+ARG OPENMPI_URL=https://download.open-mpi.org/release/open-mpi/v%OMPI_BASE%/${OPENMPI_TARNAME}.tar.gz
+RUN export OMPI_BASE=${OPENMPI_VERSION%.*} &&\
+    cd /stage && curl -fSsL `echo ${OPENMPI_URL} | sed s/%OMPI_BASE%/$OMPI_BASE/` | tar xzf - &&\
+    cd ${OPENMPI_TARNAME} &&\
+    ./configure \
+        --prefix=/opt/openmpi-${OPENMPI_VERSION} \
```

Comment:
consider moving this path to a variable to be reused during ort build

Reply:
Which path? /opt/openmpi-${OPENMPI_VERSION}? I'm not a fan of setting MPI_PATH=/opt/openmpi-${OPENMPI_VERSION} etc. It makes the actual path less clear when reading. And there's little benefit. It's all to easy to find-and-replace if you need to change it.

Reply:
Eh, I'll change it anyway.